### PR TITLE
Adding autopilot rebalance test for longevity

### PR DESCRIPTION
**Pull request by lsrinivas-pure**
**What this PR does / why we need it**:
Adding longevity test for cresting rebalance autopilot rule and validating the rule trigger
- [x] Create autopilot rebalnce rule
- [x] Validate autopilot events
- [x] Validate the pool 

**Which issue(s) this PR fixes** (optional)
Closes #PTX-13130

**Special notes for your reviewer**:

```
STEP: Create autopilot rule
INFO[2022-10-12 14:33:45] Creating autopilot rule ; {TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:pool-rebalance-absolute-prov-120-used-70 GenerateName: Namespace: SelfLink: UID: ResourceVersion: Generation:0 CreationTimestamp:0001-01-01 00:00:00 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName: ManagedFields:[]} Spec:{Weight:0 PollInterval:0 Enforcement: Selector:{LabelSelector:{MatchLabels:map[autopilot:rebalance] MatchExpressions:[]}} NamespaceSelector:{LabelSelector:{MatchLabels:map[] MatchExpressions:[]}} Conditions:{Expressions:[0xc0017440a0 0xc0017440f0] For:0 RequiredMatches:0 Type: Provider:} Actions:[0xc0013a4108] ActionsCoolDownPeriod:3600}}
INFO[2022-10-12 14:33:45] Adding label map[autopilot:rebalance] to the node torpedo-long-fa-sample-test-70-6
INFO[2022-10-12 14:33:46] Added label on torpedo-long-fa-sample-test-70-6 node: autopilot=rebalance
INFO[2022-10-12 14:33:46] Created autopilot rule: {
	"metadata": {
		"name": "pool-rebalance-absolute-prov-120-used-70",
		"uid": "ac811034-6126-47fc-998b-941ee6d7a4d9",
		"resourceVersion": "303597",
		"generation": 1,
		"creationTimestamp": "2022-10-12T09:03:46Z",
		"labels": {
			"app": "cassandra",
			"creator": "torpedo"
		},
		"managedFields": [
			{
				"manager": "basic.test",
				"operation": "Update",
				"apiVersion": "autopilot.libopenstorage.org/v1alpha1",
				"time": "2022-10-12T09:03:46Z",
				"fieldsType": "FieldsV1",
				"fieldsV1": {
					"f:metadata": {
						"f:labels": {
							".": {},
							"f:app": {},
							"f:creator": {}
						}
					},
					"f:spec": {
						".": {},
						"f:actions": {},
						"f:conditions": {
							".": {},
							"f:expressions": {}
						},
						"f:namespaceSelector": {},
						"f:selector": {
							".": {},
							"f:matchLabels": {
								".": {},
								"f:autopilot": {}
							}
						}
					}
				}
			}
		]
	},
	"spec": {
		"selector": {
			"matchLabels": {
				"autopilot": "rebalance"
			}
		},
		"namespaceSelector": {},
		"conditions": {
			"expressions": [
				{
					"key": "100 * (px_pool_stats_provisioned_bytes/ on (pool) px_pool_stats_total_bytes)",
					"operator": "Gt",
					"values": [
						"120"
					]
				},
				{
					"key": "100 * (px_pool_stats_used_bytes/ on (pool) px_pool_stats_total_bytes)",
					"operator": "Gt",
					"values": [
						"70"
					]
				}
			]
		},
		"actions": [
			{
				"name": "openstorage.io.action.storagepool/rebalance",
				"params": null
			}
		]
	}
}

```

```
STEP: validate the  autopilot events
INFO[2022-10-12 15:02:16] autopilot rule event reason : Transition and message: rule: pool-rebalance-absolute-prov-120-used-70:b163f313-e5ea-404a-af1b-f22cdf724743 transition from  => Initializing
INFO[2022-10-12 15:02:16] autopilot rule event reason : Transition and message: rule: pool-rebalance-absolute-prov-120-used-70:b163f313-e5ea-404a-af1b-f22cdf724743 transition from Initializing => Normal
INFO[2022-10-12 15:02:16] autopilot rule event reason : Transition and message: rule: pool-rebalance-absolute-prov-120-used-70:67ee64f9-7ca9-4317-95b5-5ee94ce50552 transition from  => Initializing
INFO[2022-10-12 15:02:16] autopilot rule event reason : Transition and message: rule: pool-rebalance-absolute-prov-120-used-70:67ee64f9-7ca9-4317-95b5-5ee94ce50552 transition from Initializing => Normal
```

```
STEP: validate Px on the rebalanced node
INFO[2022-10-13 11:54:53] Validating PX on node : torpedo-long-fa-sample-test-70-0
DEBU[2022-10-13 11:54:53] waiting for PX node to be up: torpedo-long-fa-sample-test-70-0
DEBU[2022-10-13 11:54:53] Getting node info for node: [torpedo-long-fa-sample-test-70-0]
DEBU[2022-10-13 11:54:54] checking PX status on node: torpedo-long-fa-sample-test-70-0
DEBU[2022-10-13 11:54:54] Finding the debug pod to run command on node torpedo-long-fa-sample-test-70-0
DEBU[2022-10-13 11:54:54] Running command on pod debug-7khhp [[nsenter --mount=/hostproc/1/ns/mnt /bin/bash -c which pxctl]]
DEBU[2022-10-13 11:54:57] Finding the debug pod to run command on node torpedo-long-fa-sample-test-70-0
DEBU[2022-10-13 11:54:57] Running command on pod debug-7khhp [[nsenter --mount=/hostproc/1/ns/mnt /bin/bash -c sudo /usr/local/bin/pxctl -j status]]
INFO[2022-10-13 11:55:01] px on node: torpedo-long-fa-sample-test-70-0 is now up. status: STATUS_OK
DEBU[2022-10-13 11:55:01] checking if PX pod is up on node: torpedo-long-fa-sample-test-70-0
DEBU[2022-10-13 11:55:02] px is fully operational on node: torpedo-long-fa-sample-test-70-0
INFO[2022-10-13 11:55:02] Job 1036049641065142325 is in DONE state
INFO[2022-10-13 11:55:02] Trigger Function completed for [autopilotRebalance]
```

### CODE REVIEW

File: tests/testTriggers.go [hunk 1]
```diff
@@ -177,6 +179,9 @@ var eventRing *ring.Ring
 //decommissionedNode for rejoin test
 var decommissionedNode = node.Node{}
 
+//node with auti pilot rule enabled
```
Comment (jainsmit) [inline tests/testTriggers.go:182]: nit: Add space after `//` and `node with auti pilot rule enabled`. Can we change the variable name to `autoPilotLabelNode`

File: tests/testTriggers.go [hunk 2]
```diff
@@ -3735,6 +3763,178 @@ func TriggerPoolAddDiskAndReboot(contexts *[]*scheduler.Context, recordChan *cha
...
+		Start:   time.Now().Format(time.RFC1123),
+		Outcome: []error{},
+	}
+
+	defer func() {
+		event.End = time.Now().Format(time.RFC1123)
+		*recordChan <- event
+	}()
+	poolLabel := map[string]string{"autopilot": "rebalance"}
+	apRule := aututils.PoolRuleRebalanceAbsolute(120, 70, false)
```
Comment (jainsmit) [inline tests/testTriggers.go:3782]: Can we avoid hardcoded value here?

File: tests/testTriggers.go [hunk 3]
```diff
@@ -3735,6 +3763,178 @@ func TriggerPoolAddDiskAndReboot(contexts *[]*scheduler.Context, recordChan *cha
...
+	apRule.Spec.ActionsCoolDownPeriod = int64(3600)
```
Comment (jainsmit) [inline tests/testTriggers.go:3783]: Same

File: tests/testTriggers.go [hunk 4]
```diff
@@ -3735,6 +3763,178 @@ func TriggerPoolAddDiskAndReboot(contexts *[]*scheduler.Context, recordChan *cha
...
+							logrus.Infof("Job %v is in Running state", job.GetId())
+
+							currentDone, total := getReblanceWorkSummary(jobResponse)
+							//checking for rebalance progress
+							for currentDone < total && previousDone < currentDone {
+								time.Sleep(2 * time.Minute)
+								logrus.Infof("Waiting for job %v to complete current state: %v, checking again in 2 minutes", job.GetId(), jobState)
+								jobResponse, err = Inst().V.GetRebalanceJobStatus(job.GetId())
+								UpdateOutcome(event, err)
+								if err == nil {
```
Comment (jainsmit) [inline tests/testTriggers.go:3891]: we can use `if err != nil then break` and remove the else part

File: tests/testTriggers.go [hunk 5]
```diff
@@ -3633,8 +3643,16 @@ func TriggerPoolResizeDiskAndReboot(contexts *[]*scheduler.Context, recordChan *
 			logrus.Error(err)
 			UpdateOutcome(event, err)
 		}
+		var poolToBeResized *opsapi.StoragePool
 		if len(poolsToBeResized) > 0 {
-			poolToBeResized := poolsToBeResized[0]
+			for _, pool := range poolsToBeResized {
+
```
Comment (jainsmit) [inline tests/testTriggers.go:3649]: Delete this line

File: tests/testTriggers.go [hunk 6]
```diff
@@ -3633,8 +3643,16 @@ func TriggerPoolResizeDiskAndReboot(contexts *[]*scheduler.Context, recordChan *
...
+				if !isPoolRebalanceEnabled(pool.Uuid) {
+					poolToBeResized = pool
+				}
+
```
Comment (jainsmit) [inline tests/testTriggers.go:3653]: same

File: tests/testTriggers.go [hunk 7]
```diff
@@ -3719,8 +3740,15 @@ func TriggerPoolAddDiskAndReboot(contexts *[]*scheduler.Context, recordChan *cha
...
 			UpdateOutcome(event, err)
 		}
+		var poolToBeResized *opsapi.StoragePool
 		if len(poolsToBeResized) > 0 {
-			poolToBeResized := poolsToBeResized[0]
+			for _, pool := range poolsToBeResized {
+				if !isPoolRebalanceEnabled(pool.Uuid) {
+					poolToBeResized = pool
+				}
+
```
Comment (jainsmit) [inline tests/testTriggers.go:3749]: same

File: tests/testTriggers.go (hunk 1 above)
Comment (lsrinivas-pure) [reply tests/testTriggers.go:182]: fixed

File: tests/testTriggers.go (hunk 2 above)
Comment (lsrinivas-pure) [reply tests/testTriggers.go:3782]: these are standard values mentioned in docs. I dont think this needs to be parameterized

File: tests/testTriggers.go (hunk 3 above)
Comment (lsrinivas-pure) [reply tests/testTriggers.go:3783]: fixed

File: tests/testTriggers.go (hunk 4 above)
Comment (lsrinivas-pure) [reply tests/testTriggers.go:3891]: fixed

File: tests/testTriggers.go (hunk 5 above)
Comment (lsrinivas-pure) [reply tests/testTriggers.go:3649]: fixed

File: tests/testTriggers.go (hunk 7 above)
Comment (lsrinivas-pure) [reply tests/testTriggers.go:3749]: fixed

File: tests/testTriggers.go (hunk 6 above)
Comment (lsrinivas-pure) [reply tests/testTriggers.go]: fixed

File: tests/testTriggers.go (hunk 2 above)
Comment (jainsmit) [reply tests/testTriggers.go:3912]: Can we define these as constant and use them ?

**Review by jainsmit — APPROVED** [review]

**Pull request merged**